go/ct/gen/interval_solver_test.go

```diff
@@ -0,0 +1,326 @@
+				s.AddLowerBoundary(300)
+			},
+			want: "X ∈ [300..400]",
+		},
+		"upper-boundary-constraint": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.AddUpperBoundary(300)
+			},
+			want: "X ∈ [200..300]",
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			solver := NewIntervalSolver[int32](200, 400)
+			test.setup(solver)
+			got := solver.String()
+			if got != test.want {
+				t.Errorf("unexpected result, got %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+func TestIntervalSolver_Exclude(t *testing.T) {
+	const N = 20
+	// remove interval [c..d] from [a..b]
+	for a := 0; a < N; a++ {
+		for b := 0; b < N; b++ {
+			for c := 0; c < N; c++ {
+				for d := 0; d < N; d++ {
+					solver := NewIntervalSolver(uint32(a), uint32(b))
+					solver.Exclude(uint32(c), uint32(d))
+					for i := 0; i < N; i++ {
+						inAtoB := a <= i && i <= b
+						inCtoD := c <= i && i <= d
+						want := inAtoB && !inCtoD
+						got := solver.Contains(uint32(i))
+						if got != want {
+							t.Fatalf("unexpected result for %d ∈ [%d..%d] \\ [%d..%d]: got %v, want %v", i, a, b, c, d, got, want)
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func TestIntervalSolver_EmptyInterval(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 100)
+	_, err := solver.Generate(rand.New())
+	if err != ErrUnsatisfiable {
+		t.Errorf("failed to generate point in empty interval: %v", err)
+	}
+}
+
+func TestIntervalSolver_ExcludeEmptyIntervalHasNoEffect(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	clone := solver.Clone()
+	clone.Exclude(300, 250)
+	if !solver.Equals(clone) {
+		t.Fatalf("empty interval should not be modified by adding empty interval")
+	}
+}
+
+func TestIntervalSolver_TestIntervals(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	solver.Exclude(250, 300)
+
+	rnd := rand.New()
+	for i := 0; i < 100; i++ {
+		res, err := solver.Generate(rnd)
+		if err != nil {
+			t.Fatalf("error solving intervals: %v", err)
+		}
+		if res < 200 || res > 400 || (res >= 250 && res <= 300) {
+			t.Fatalf("produced unexpected result for condition %v: %d", solver, res)
+		}
+	}
+}
+
+func TestIntervalSolver_IsSatisfiable(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	if !solver.IsSatisfiable() {
+		t.Fatalf("interval should be solvable: %s", solver.String())
+	}
+	solver.intervals = nil
+	if solver.IsSatisfiable() {
+		t.Fatalf("empty interval should be unsolvable")
+	}
+}
+
+func TestIntervalSolver_AddBoundariesInEmptySolverHasNoEffect(t *testing.T) {
+
+	tests := map[string]func(*IntervalSolver[int32]){
+		"lower":    func(s *IntervalSolver[int32]) { s.AddLowerBoundary(300) },
+		"upper":    func(s *IntervalSolver[int32]) { s.AddUpperBoundary(300) },
+		"equality": func(s *IntervalSolver[int32]) { s.AddEqualityConstraint(300) },
+	}
+
+	for name, setup := range tests {
+		t.Run(name, func(t *testing.T) {
+
+			solver := NewIntervalSolver[int32](200, 400)
+			solver.intervals = nil
```

Comment:
I would prefer creating the empty interval through public API.
```suggestion
			solver := NewIntervalSolver[int32](400, 200)
```